File: t-tasks/src/main/java/com/teo/ttasks/receivers/TaskNotificationReceiver.java
package com.teo.ttasks.receivers;

import android.app.Notification;
import android.app.NotificationManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.widget.Toast;

import com.teo.ttasks.TTasksApp;
import com.teo.ttasks.data.remote.TasksHelper;

import javax.inject.Inject;

import io.realm.Realm;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import timber.log.Timber;

public class TaskNotificationReceiver extends BroadcastReceiver {

    public static final String NOTIFICATION_ID = "notification-id";
    public static final String NOTIFICATION = "notification";

    public static final String TASK_ID = "taskId";

    public static final String ACTION_PUBLISH = "publish";
    public static final String ACTION_COMPLETE = "complete";
    public static final String ACTION_DELETE = "delete";

    @Inject TasksHelper tasksHelper;

    @Override
    public void onReceive(Context context, Intent intent) {
        final String action = intent.getAction();
        if (action != null) {
            NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
            String taskId = intent.getStringExtra(TASK_ID);
            int id = intent.getIntExtra(NOTIFICATION_ID, 0);
            Realm realm;
            switch (action) {
                case ACTION_PUBLISH:
                    // Display the notification
                    Notification notification = intent.getParcelableExtra(NOTIFICATION);
                    notificationManager.notify(id, notification);
                    break;
                case ACTION_COMPLETE:
                    TTasksApp.get(context).userComponent().inject(this);
                    realm = Realm.getDefaultInstance();
                    // Mark the task as completed
                    tasksHelper.updateCompletionStatus(taskId, realm)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(
                                    tTask -> {
                                        Toast.makeText(context, "Task completed", Toast.LENGTH_SHORT).show();
                                        // Update successful, update sync status
                                        realm.executeTransaction(realm1 -> tTask.setSynced(true));
                                        realm.close();
                                        notificationManager.cancel(id);
                                    },
                                    throwable -> {
                                        Timber.e(throwable.toString());
                                        Toast.makeText(context, "Task not found. This is the case if the task was deleted.", Toast.LENGTH_SHORT).show();
                                        realm.close();
                                        notificationManager.cancel(id);
                                    }
                            );
                    break;
                case ACTION_DELETE:
                    // Mark this task's reminder as dismissed
                    TTasksApp.get(context).userComponent().inject(this);
                    realm = Realm.getDefaultInstance();
                    tasksHelper.getTask(taskId, realm)
                            .first()
                            .flatMap(tTask -> {
                                if (tTask == null)
                                    return Observable.error(new Throwable("Task not found"));
                                return Observable.just(tTask);
                            })
                            .subscribe(
                                    tTask -> {
                                        realm.executeTransaction(realm1 -> tTask.setNotificationDismissed(true));
                                        realm.close();
                                        Timber.d("task marked as dismissed");
                                    },
                                    throwable -> {
                                        Timber.e(throwable.toString());
                                        realm.close();
                                    }
                            );
                    break;
            }
        }
    }
}

File: t-tasks/src/main/java/com/teo/ttasks/receivers/TaskNotificationReceiver.kt
package com.teo.ttasks.receivers

import android.app.Notification
import android.app.NotificationManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.widget.Toast
import com.teo.ttasks.data.remote.TasksHelper
import io.realm.Realm
import org.koin.core.KoinComponent
import org.koin.core.inject
import timber.log.Timber

class TaskNotificationReceiver : BroadcastReceiver(), KoinComponent {

    companion object {
        const val NOTIFICATION_ID = "notification-id"
        const val NOTIFICATION = "notification"

        const val TASK_ID = "taskId"

        const val ACTION_PUBLISH = "publish"
        const val ACTION_COMPLETE = "complete"
        const val ACTION_DELETE = "delete"
    }

    private val tasksHelper: TasksHelper by inject()

    override fun onReceive(context: Context, intent: Intent?) {
        val action = intent?.action ?: return
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val taskId = intent.getStringExtra(TASK_ID)
        val id = intent.getIntExtra(NOTIFICATION_ID, 0)
        when (action) {
            ACTION_PUBLISH -> {
                // Display the notification
                val notification = intent.getParcelableExtra<Notification>(NOTIFICATION)
                notificationManager.notify(id, notification)
            }
            ACTION_COMPLETE -> {
                val realm = Realm.getDefaultInstance()
                // Mark the task as completed
                if (tasksHelper.updateCompletionStatus(taskId!!, realm)) {
                    Toast.makeText(context, "Task completed", Toast.LENGTH_SHORT).show()
                    notificationManager.cancel(id)
                } else {
                    Toast.makeText(context, "Task not found. This is the case if the task was deleted.", Toast.LENGTH_SHORT).show()
                }
                notificationManager.cancel(id)
                realm.close()
            }
            ACTION_DELETE -> {
                // Mark this task's reminder as dismissed
                val realm = Realm.getDefaultInstance()
                tasksHelper.getTaskAsSingle(taskId!!, realm)
                        .subscribe(
                                { task ->
                                    realm.executeTransaction { task.notificationDismissed = true }
                                    realm.close()
                                    Timber.v("Dismissed task %s", taskId)
                                },
                                { throwable ->
                                    Timber.e(throwable, "Error while dismissing notification for %s", taskId)
                                    realm.close()
                                }
                        )
            }
            else -> {
                // do nothing
            }
        }
    }
}
